Default observer and fake-quant for backends

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26627 Default observer and fake-quant for backends**
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* #26624 Quantization aware training: Freeze batch norm support
* #26623 Per channel fake quant

Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

raghuramank10000 committed:
Default observer and fake-quant for backends



Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/QConfig.py:
```diff
@@ -63,3 +65,25 @@ def __new__(cls, weight):
 default_qat_qconfig = QConfig(activation=default_fake_quant,
                               weight=default_weight_fake_quant)
+
+def get_default_qconfig(backend='fbgemm'):
```
technically - it's not even fbgemm, but rather fbgemm on per-CooperLake x86 :-\ . But I don't know how to make it nicer.

raghuramank100 replied:
I agree, perhaps just call it server and have a comment stating that you need to know about the hardware platform to ensure you are picking a good default?

raghuramank100 replied:
Actually with mkl-dnn not sure how to improve on this, so leaving this as is for now

dzhulgakov commented on torch/quantization/QConfig.py:
```diff
@@ -63,3 +65,25 @@ def __new__(cls, weight):
+def get_default_qconfig(backend='fbgemm'):
+    if backend == 'fbgemm':
+        qconfig = QConfig(activation=HistogramObserver.with_args(reduce_range=True),
+                          weight=default_per_channel_weight_observer)
```
so we do per-channel only for fbgemm, not for mobile?

raghuramank100 replied:
Mobile does not support per-channel kernels currently. If and when QNNPACK supports it, we can change that.

dzhulgakov commented on torch/quantization/QConfig.py:
```diff
@@ -63,3 +65,25 @@ def __new__(cls, weight):
+
+def get_default_qat_qconfig(backend='fbgemm'):
+    if backend == 'fbgemm':
+        qconfig = QConfig(activation=FakeQuantize.with_args(observer=default_l2_observer(reduce_range = True),
```
I think this will create the same instance of observer for each FakeQuant - you probably need double with_args, but it gets ugly...

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Feel free to land this PR before the others (reorder the stack) as it's useful

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Default observer and fake-quant for backends"


Differential Revision: [D17518194](https://our.internmc.facebook.com/intern/diff/D17518194/)

[ghstack-poisoned]
